### PR TITLE
[key-manager] condition test and action handler

### PR DESCRIPTION
**Pull request by davidiw**
This stack of commits builds a testing framework for secure and then uses it to evaluate key-manager handling. The key-manager framework is effectively around a set of conditions to evaluate to determine an action as well as an action handler. This PR adds that as well as a thorough test on it.

### CODE REVIEW

File: secure/key-manager/src/tests.rs [hunk 1]
```diff
@@ -100,30 +106,31 @@ impl Node {
...
         }
     }
 
-    // TODO(davidiw): Notify external upon reconfigure and update epoch via ledger_info and also add
-    // next validator set if appropriate...
     fn execute_and_commit(&self, mut block: Vec<Transaction>) {
+        // 1) Update the clock for potential reconfigurations
+        self.time.increment();
+        // Clock is supposed to be in microseconds
+        let clock = self.time.now() * 1_000_000;
```
Comment (zekun000) [inline secure/key-manager/src/tests.rs]: you probably can put the clock in the LI to make it more real

File: secure/storage/src/on_disk.rs [hunk 2]
```diff
@@ -55,14 +71,9 @@ impl OnDiskStorage {
...
     }
-
-    /// Public convenience function to return a new OnDiskStorage based Storage.
-    pub fn new_storage(path_buf: PathBuf) -> Box<dyn Storage> {
-        Box::new(OnDiskStorage::new(path_buf))
-    }
 }
 
-impl KVStorage for OnDiskStorage {
+impl<T: Send + Sync + TimeService> KVStorage for OnDiskStorageInternal<T> {
```
Comment (zekun000) [inline secure/storage/src/on_disk.rs]: from past experience, put Send + Sync in your trait saves more time typing

File: secure/key-manager/src/lib.rs [hunk 3]
```diff
@@ -160,15 +173,62 @@ where
...
+        if let Err(Error::ConfigInfoKeyMismatch(..)) = self.compare_info_to_config() {
+            // For now, just assume this is correct, but this needs to compare libra_timestamp with
+            // the current time to ensure that progress is being made. And if not flag an error.
+            return Ok(Action::NoAction);
+        }
+
+        let last_rotation = self.last_rotation()?;
+
+        // If this is inconsistent, then the transaction either failed or was never submitted.
+        if let Err(Error::ConfigStorageKeyMismatch(..)) = self.compare_storage_to_config() {
```
Comment (zekun000) [inline secure/key-manager/src/lib.rs]: the function name is deceptive, I was wondering why we're comparing twice

File: secure/storage/src/on_disk.rs (hunk 2 above)
Comment (davidiw) [reply secure/storage/src/on_disk.rs]: My hope is to eliminate the Send + Sync at some point in the future... actually I wonder if it is even necessary any more...

File: secure/key-manager/src/tests.rs (hunk 1 above)
Comment (davidiw) [reply secure/key-manager/src/tests.rs]: Hmm.... lemme think about that.

**Pull request closed without merging**